In [3]:
import pandas as pd
import numpy as np
df = pd.read_csv(r'C:\Users\91704\Desktop\Brentwood_TakeHome\Data_raw\online_retail_data.csv')
df.head()

customer_id  order_date  product_id  category_id       category_name  \
0        13542  12/17/2024         784           10         Electronics   
1        23188    6/1/2024         682           50   Sports & Outdoors   
2        55098    2/4/2025         684           50   Sports & Outdoors   
3        65208  10/28/2024         204           40  Books & Stationery   
4        63872   5/10/2024         202           20             Fashion   

  product_name  quantity   price payment_method            city  review_score  \
0   Smartphone         2  373.36    Credit Card  New Oliviaberg           1.0   
1  Soccer Ball         5  299.34    Credit Card    Port Matthew           NaN   
2         Tent         5   23.00    Credit Card      West Sarah           5.0   
3   Story Book         2  230.11  Bank Transfer  Hernandezburgh           5.0   
4        Skirt         4  176.72    Credit Card    Jenkinshaven           1.0   

  gender  age  
0      F   56  
1      M   59  
2      F   64  
3      M   34  
4      F   33

In [5]:
df.info()
df.describe(include="all")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   customer_id     1000 non-null   int64  
 1   order_date      1000 non-null   object 
 2   product_id      1000 non-null   int64  
 3   category_id     1000 non-null   int64  
 4   category_name   1000 non-null   object 
 5   product_name    1000 non-null   object 
 6   quantity        1000 non-null   int64  
 7   price           1000 non-null   float64
 8   payment_method  1000 non-null   object 
 9   city            1000 non-null   object 
 10  review_score    799 non-null    float64
 11  gender          897 non-null    object 
 12  age             1000 non-null   int64  
dtypes: float64(2), int64(5), object(6)
memory usage: 101.7+ KB


customer_id order_date   product_id  category_id      category_name  \
count    1000.000000       1000  1000.000000  1000.000000               1000   
unique           NaN        342          NaN          NaN                  5   
top              NaN  8/12/2024          NaN          NaN  Sports & Outdoors   
freq             NaN         10          NaN          NaN                211   
mean    55490.723000        NaN   540.726000    30.030000                NaN   
std     25910.185857        NaN   261.737704    14.370303                NaN   
min     10201.000000        NaN   100.000000    10.000000                NaN   
25%     33857.000000        NaN   311.750000    20.000000                NaN   
50%     54619.500000        NaN   542.500000    30.000000                NaN   
75%     77848.500000        NaN   770.750000    40.000000                NaN   
max     99923.000000        NaN   995.000000    50.000000                NaN   

       product_name     quantity        price    payment_method          city  \
count          1000  1000.000000  1000.000000              1000          1000   
unique           25          NaN          NaN                 3           962   
top        Yoga Mat          NaN          NaN  Cash on Delivery  East William   
freq             51          NaN          NaN               374             3   
mean            NaN     2.947000   251.850660               NaN           NaN   
std             NaN     1.413573   139.194688               NaN           NaN   
min             NaN     1.000000    10.720000               NaN           NaN   
25%             NaN     2.000000   128.525000               NaN           NaN   
50%             NaN     3.000000   250.220000               NaN           NaN   
75%             NaN     4.000000   366.467500               NaN           NaN   
max             NaN     5.000000   499.500000               NaN           NaN   

        review_score gender          age  
count     799.000000    897  1000.000000  
unique           NaN      2          NaN  
top              NaN      M          NaN  
freq             NaN    457          NaN  
mean        3.992491    NaN    46.382000  
std         1.239469    NaN    16.569992  
min         1.000000    NaN    18.000000  
25%         3.000000    NaN    32.000000  
50%         4.000000    NaN    47.000000  
75%         5.000000    NaN    61.000000  
max         5.000000    NaN    75.000000

In [4]:
df["order_date"] = pd.to_datetime(df["order_date"], errors="coerce")
df["order_date"].isna().sum()
df.dtypes


customer_id                int64
order_date        datetime64[ns]
product_id                 int64
category_id                int64
category_name             object
product_name              object
quantity                   int64
price                    float64
payment_method            object
city                      object
review_score             float64
gender                    object
age                        int64
dtype: object

In [5]:
df["revenue"] = df["price"] * df["quantity"]
df[["price", "quantity", "revenue"]].head()



price  quantity  revenue
0  373.36         2   746.72
1  299.34         5  1496.70
2   23.00         5   115.00
3  230.11         2   460.22
4  176.72         4   706.88

In [6]:
(price_issues := (df["price"] <= 0).sum())
(quantity_issues := (df["quantity"] <= 0).sum())
(revenue_issues := (df["revenue"] <= 0).sum())

(review_low := (df["review_score"] < 0).sum())
(review_high := (df["review_score"] > 5).sum())

(age_low := (df["age"] < 10).sum())
(age_high := (df["age"] > 100).sum())

price_issues, quantity_issues, revenue_issues, review_low, review_high, age_low, age_high


(0, 0, 0, 0, 0, 0, 0)

In [7]:

df.duplicated().sum() #checks if entire row is duplicated
df.duplicated(subset=["customer_id", "order_date", "product_id"]).sum()     #Checks specifically for customer id-orderdate- product id is repeated.


0

In [8]:
df.isna().sum().sort_values(ascending=False)


review_score      201
gender            103
customer_id         0
order_date          0
product_id          0
category_id         0
category_name       0
product_name        0
quantity            0
price               0
payment_method      0
city                0
age                 0
revenue             0
dtype: int64

201 missing review score, but i will be leaving them as-is. Can take median in analysis if needed. 
103 Gender missing values will be filled be "Unknown" remark. Will be needed for reporting purposes. 

In [9]:
df["gender"] = df["gender"].fillna("Unknown")
df.isna().sum()


customer_id         0
order_date          0
product_id          0
category_id         0
category_name       0
product_name        0
quantity            0
price               0
payment_method      0
city                0
review_score      201
gender              0
age                 0
revenue             0
dtype: int64

In [10]:
total_revenue = df["revenue"].sum()
total_orders = len(df)   # Since i don't have order_id in raw file, each row = one order
total_customers = df["customer_id"].nunique()

aov = total_revenue / total_orders

profit = total_revenue * 0.30   # 30% margin as per assumption
profit_margin = profit / total_revenue

total_revenue, total_orders, total_customers, aov, profit, profit_margin


(737326.8800000001, 1000, 1000, 737.3268800000001, 221198.06400000004, 0.3)

Total Revenue: $737,326.88 // Assuming the currency is USD based on location names

Total Orders: 1,000

Total Customers: 1,000

AOV: $737.33

Profit (30%): $221,198.06

Profit Margin: 30%

In [11]:
# Create monthly period column
df["order_month"] = df["order_date"].dt.to_period("M").dt.to_timestamp()

# Aggregate by month
monthly_perf = df.groupby("order_month").agg(
    monthly_revenue=("revenue", "sum"),
    monthly_orders=("customer_id", "count")
).reset_index()

# Month-over-Month Growth
monthly_perf["mom_growth_pct"] = monthly_perf["monthly_revenue"].pct_change() * 100

monthly_perf


order_month  monthly_revenue  monthly_orders  mom_growth_pct
0   2024-03-01         30620.98              40             NaN
1   2024-04-01         50375.40              79       64.512697
2   2024-05-01         61448.18              92       21.980530
3   2024-06-01         59127.40              75       -3.776808
4   2024-07-01         57939.86              90       -2.008443
5   2024-08-01         79070.12              90       36.469298
6   2024-09-01         69323.09              90      -12.327071
7   2024-10-01         55328.98              73      -20.186795
8   2024-11-01         55404.74              75        0.136926
9   2024-12-01         78432.94              97       41.563592
10  2025-01-01         60210.83              89      -23.232726
11  2025-02-01         48826.04              67      -18.908210
12  2025-03-01         31218.32              43      -36.062150

In [12]:
city_perf = df.groupby("city").agg(
    revenue=("revenue", "sum"),
    orders=("customer_id", "count"),
    customers=("customer_id", "nunique")
).reset_index()

city_perf["aov"] = city_perf["revenue"] / city_perf["orders"]

city_perf = city_perf.sort_values(by="revenue", ascending=False)
city_perf


city  revenue  orders  customers          aov
672  Port Melissaborough  3941.29       2          2  1970.645000
597        Patriciaville  3324.10       2          2  1662.050000
298       Johnsonborough  3045.09       2          2  1522.545000
189         East William  3011.36       3          3  1003.786667
140     East Christopher  2919.20       2          2  1459.600000
..                   ...      ...     ...        ...          ...
225        Gonzalezshire    27.65       1          1    27.650000
660   Port Kimberlymouth    24.00       1          1    24.000000
955           Williamton    22.21       1          1    22.210000
554          North Diane    21.44       1          1    21.440000
302         Jonathanstad    20.84       1          1    20.840000

[962 rows x 5 columns]

Category that makes more money 

In [13]:
category_perf = df.groupby("category_name").agg(
    revenue=("revenue", "sum"),
    orders=("customer_id", "count"),
    customers=("customer_id", "nunique")
).reset_index()

category_perf["aov"] = category_perf["revenue"] / category_perf["orders"]

category_perf = category_perf.sort_values(by="revenue", ascending=False)
category_perf


category_name    revenue  orders  customers         aov
1         Electronics  166510.34     207        207  804.397778
4   Sports & Outdoors  154346.26     211        211  731.498863
0  Books & Stationery  143215.52     193        193  742.049326
3       Home & Living  138540.15     191        191  725.341099
2             Fashion  134714.61     198        198  680.376818

In [14]:
gender_perf = df.groupby("gender").agg(
    revenue=("revenue", "sum"),
    orders=("customer_id", "count"),
    customers=("customer_id", "nunique")
).reset_index()

gender_perf["aov"] = gender_perf["revenue"] / gender_perf["orders"]

gender_perf = gender_perf.sort_values(by="revenue", ascending=False)
gender_perf


gender    revenue  orders  customers         aov
1        M  333894.19     457        457  730.621860
0        F  331753.65     440        440  753.985568
2  Unknown   71679.04     103        103  695.913010

In [15]:
#finding the age group which spends more

bins = [0, 25, 35, 45, 60, 120]
labels = ["18–25", "26–35", "36–45", "46–60", "60+"]

df["age_bucket"] = pd.cut(df["age"], bins=bins, labels=labels, right=True)

age_perf = df.groupby("age_bucket").agg(
    revenue=("revenue", "sum"),
    orders=("customer_id", "count"),
    customers=("customer_id", "nunique")
).reset_index()

age_perf["aov"] = age_perf["revenue"] / age_perf["orders"]

age_perf
df.to_csv("sales_clean.csv", index=False)


C:\Users\91704\AppData\Local\Temp\ipykernel_22168\1796567283.py:8: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  age_perf = df.groupby("age_bucket").agg(


In [16]:
df.to_csv("sales_clean.csv", index=False)
